### DISCUSSION TITLE
How does Homebrew treat aliases with a version tag?

### DISCUSSION THREAD

**Discussion by ymleung314** [Homebrew/discussions#5101] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: cfac516ee045d869be34598152e80a11ef2befe7
Last commit: 4 days ago
Core tap JSON: 01 Feb 06:36 UTC
Core cask tap JSON: 01 Feb 06:36 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

Lately, I was installing OpenJDK (https://formulae.brew.sh/formula/openjdk). I can see that only LTS versions, i.e. 8, 11 and 17, are "versioned" as separate casks. The current version 21 which is also the latest LTS is somehow an alias to the regular `openjdk` cask. When I ran `brew install openjdk@21`, all logs suggested that I am being redirected to the plain `openjdk`. I wonder if OpenJDK 22 is released and I run `brew upgrade`. Will my installed `openjdk` be upgraded to 22 or be kept separately as `openjdk@21`?

**Comment by ZhongRuoyu** [comment]
The latest version of `openjdk` always has an `openjdk@x` alias; the former is the canonical name so you see that instead of the versioned one. We only keep previous LTS releases (8, 11, 17) as versioned formulae.

When OpenJDK 22 is released, `openjdk` will be upgraded to 22, and a separate versioned formula `openjdk@21` will be created.

**Reply by ymleung314** [reply]
Assume that I want OpenJDK 21 only. Once OpenJDK 22 is released, I will need to manually uninstall `openjdk`, wait for `openjdk@21` to be converted into a formula instead of an alias first, then install `openjdk@21`. Is that correct?

**Reply by SMillerDev** [reply]
You can also install `openjdk@21` and `openjdk` alongside, but yes. You would need to install the older version separately.

**Comment by ymleung314** [comment]
Although not sure if this is the place to ask for it, can it be done like
how it is done with Python? I know that Python casks are created for every
minor versions while OpenJDK casks are created for LTS versions only though.

**Reply by SMillerDev** [reply]
Python minor versions are LTS versions.

**Reply by ymleung314** [reply]
My point is that we may do it like how `python` is an alias to a versioned cask instead of like how `openjdk@21` is an alias to `openjdk`. But I can already think of an issue in this. Python LTS versions are "continuous" (3.7, 3.8, 3.9, etc) where each minor version has its own versioned cask and the `python` alias can just reference one. But for OpenJDK, the LTS versions are "discrete" (8, 11, 17, 21, etc) and only they have versioned casks. Thus, you cannot just reference `openjdk` to, for example, `openjdk@22`.
